### DISCUSSION TITLE
Attribute API & Images Upload API

### DISCUSSION THREAD

**Discussion by JanikHalder** [swellstores/community#3] [Help] [answered]
Hi there,
I have two issues. 
Is it possible to upload Images into Swell and save them inside? 
I have tried using: 
const result = await swell.post('/products', {
        name: name,
        active: true,
        sku: sku,
        id: id,
        atro_id: id,
        category_id: category_id,
        description: description,
        attributes: attributes,
        images: images,
    });
Where I am passing images as mentioned in the old forum:
images = [{
 data: { $base64: imageDataInBase64 }
}
];

But this doesn't work.

My second issue is that I don't get the product attributes through the API that I create through the API. If i call it inside Swell I get everything as expected but if i call it in my storefront I only get an empty object.

<img width="948" alt="Bildschirm­foto 2022-11-24 um 09 48 56" src="https://user-images.githubusercontent.com/95503532/203735671-793c9276-f401-4a4d-92cc-244dfb189ea9.png">

<img width="948" alt="Bildschirm­foto 2022-11-24 um 09 49 43" src="https://user-images.githubusercontent.com/95503532/203735882-2ef912d7-5dd0-4cf0-89ee-a7cf3c9a3fba.png">

Is this a bug or have I missed something as I am creating the products through the API call.

This is how I pass the product attributes:

attributes: {
  ef000007_ec001500: 'blau [EV000080]',
  ef000008: 12,
  ef000040: 111,
  ef000279_ec001500: 'elektronischer Trafo [EV000681]',
  ef000316_ec001500: 'Schraubanschluss [EV000415]',
  ef001438: 130
}

**Comment by logeshswell** [comment, marked as answer]
@JanikHalder Please find the steps to upload images in Swell and associate to a product or any of the swell components.
1. Upload images to Swell files directory.
```ts
const swell = require('swell-node').init('store-id', 'secret-key');
let fileData = {
  filename: 'filename.png',
  content_type: 'image/png',
  data: {
    $binary: '<imageDataInBase64>'
  }
}
const fileResponse = await swell.post(`/:files`, fileData);
```

2. Associate image file to a product.
```ts
await swell.put('/products/<product_id>', {
  images: [
    { file: fileResponse }
  ]
});
```

**Reply by JanikHalder** [reply]
Thank you I will try it.

**Reply by jezzzm** [reply]
hey @logeshswell, is there a way to in one step pass an external url to the `:files` endpoint and have it downloaded/reuploaded to swell file store? similar to the way i guess it is done behind the scenes when uploading JSON via admin ui? https://developers.swell.is/guides/migrate-from-shopify#products

**Comment by logeshswell** [comment]
@JanikHalder Regarding the attribute API, the attribute will be shown in storefront API (or swell-js) only if it matches the below conditions:
1. Attribute id (eg: `ef000008`) must be present in the product attributes list (https://store-id.swell.store/admin/attributes/ef000008)
2. `Visible` toggle must be enabled for the attribute.
3. Attribute must be linked to the product that you are fetching.

On a side note, you don't need to expand by `attributes` to show the attributes for a product.

**Reply by JanikHalder** [reply]
Hi @logeshswell,
thank you for your answer. The problem was that there is a bug with the attributes,that is why the attributes got not linked to the product, I mentioned that in another post and already got the answer that this will be solved.

**Reply by logeshswell** [reply]
Yes, I am on track of that attribute pagination [issue](https://github.com/orgs/swellstores/discussions/5).
Let me keep you posted once it is resolved.
